### ISSUE TITLE
Use a queue for all requests in the request manager

### ISSUE THREAD

**Issue by provokateurin** [nextcloud/neon#1124]

**Comment by Leptopoda** [comment]
The queue can hold a record with a priority and a request.
This way we can ensure critical requests like the startup checks from the framework happen earlier.
Once images also use the request manager they can get a lower one.

I propose 4 levels:
- critical (should only be used for the startup checks inside the frsmework)
- high
- normal 
- low (secondary content like image previews)

**Comment by Leptopoda** [comment]
> Another thing we could add is that requests with the same key are deduplicated. If there is already a request in the queue with the same "target" then the second request can just listen to the result of the first one.
Ok I don't like adding all of this functionality to the RequestManager directly.
Keeping #800 in mind we should add a separate `RequestBloc` that handles the queue and priorities and can handle the global error and loading states.
The RequestBloc can then invoke the current RequestManager to do the actual request.

**Comment by Leptopoda** [comment]
I think that just a simple queue will not work.
If the connection to the server takes a while (slow server, bad connectivity) the queue will also slow down fetching the data from the cache.

Ideally the cache should not be affected and only the real network request should be queued.

**Comment by provokateurin** [comment]
Yep, though that should still be achievable using a simple queue?

**Comment by Leptopoda** [comment]
I haven't looked into it but from the description the [pool](https://pub.dev/packages/pool) package looks good.
Sounds like it would also handle timeouts.

**Comment by provokateurin** [comment]
Ideally it needs to also handle priorities :thinking:

**Comment by vauvenal5** [comment]
Regarding this, are we good to start the topic of with #3184 or do you want a more feature complete solution from the get go?

**Comment by provokateurin** [comment]
I'm fine with not doing everything in one go, however I would like to have the cached requests skip the queue, since it will improve performance quite a bit.

**Comment by vauvenal5** [comment]
I will double check behavior with skipping the queue for cached elements. Need to check how the behavior is with respect to smooth scrolling through a bigger image folder..

**Comment by vauvenal5** [comment]
Checked this in conjunction with the blurHash PR and same test environment. Looks good. I updated to PR to only wrap actual requests to the server.
